tezos/encoding/src/generator.rs

```diff
@@ -0,0 +1,689 @@
+
+macro_rules! int_type {
+    ($t:ty) => {
+        impl IntType for $t {
+            fn min_value() -> Self {
+                <$t>::MIN
+            }
+
+            fn max_value() -> Self {
+                <$t>::MAX
+            }
+
+            fn zero() -> Self {
+                0
+            }
+
+            fn one() -> Self {
+                1
+            }
+        }
+    };
+}
+
+int_type!(u8);
+int_type!(u16);
+int_type!(u32);
+int_type!(u64);
+int_type!(usize);
+
+fn decode_range_bounds<T: IntType>(bounds: impl RangeBounds<T>) -> (T, T) {
+    let min = match bounds.start_bound() {
+        Bound::Included(b) => *b,
+        Bound::Excluded(b) => *b + T::one(),
+        Bound::Unbounded => T::min_value(),
+    };
+    let max = match bounds.end_bound() {
+        Bound::Included(b) => *b,
+        Bound::Excluded(b) => *b - T::one(),
+        Bound::Unbounded => T::max_value(),
+    };
+    (min, max)
+}
+
+impl<T> IntGenerator<T>
+where
+    T: IntType,
+{
+    pub fn new(min: T, max: T, step: T, radius: T) -> Self {
+        assert!(min < max);
+        assert!(step > T::zero());
+
+        let step = cmp::min(step, max - min);
+        let radius = cmp::min(radius, step >> T::one());
+        Self {
+            min,
+            max,
+            step,
+            radius,
+            curr: min,
+        }
+    }
+}
+
+impl<T> Generator for IntGenerator<T>
+where
+    T: IntType,
+{
+    type Item = T;
+
+    fn next(&mut self) -> bool {
+        if self.curr >= self.max {
+            self.curr = self.min;
+            return false;
+        }
+        let curr_offset = (self.curr - self.min) % self.step;
+        if self.radius == curr_offset {
+            let curr_node = (self.curr - self.min) / self.step;
+            let curr_center = self.min + curr_node * self.step;
+            let mut new_curr = if curr_center > self.max - self.step {
+                self.max - self.radius
+            } else {
+                curr_center + self.step - self.radius
+            };
+            if new_curr <= self.curr {
+                new_curr = self.curr + T::one();
+            }
+            self.curr = new_curr;
+        } else {
+            self.curr = self.curr + T::one();
+        }
+        true
+    }
+
+    fn value(&self) -> Self::Item {
+        self.curr
+    }
+}
+
+/// Helper composer for producing [std::vec::Vec] generator out of lenght generator and element generator
+pub fn vec_of_items<G1, G2>(item: G1, len: G2) -> impl Generator<Item = Vec<G1::Item>>
+where
+    G1: Generator,
+    G1::Item: Clone,
+    G2: Generator<Item = usize>,
+{
+    compose(len, item, |len, item| vec![item; len])
+}
+
+/// Helper composer for producing [std::opt::Option] generator out of element generator
+pub fn option<G1, G2>(item: G1, presence: G2) -> impl Generator<Item = Option<G1::Item>>
+where
+    G1: Generator,
+    G2: Generator<Item = bool>,
+{
+    // TODO use optimized version that does not issue [None] multiple times.
```

Comment:
@akoptelov how important is this TODO? does it deserve an issue on Jira?

Reply:
This is mostly for testing purposes. So my answer is no.